File: clients/swift/Sources/AresTransport/AdminClient.swift
import Foundation

public struct AdminClient: Sendable {
    public let serverURL: String
    public init(serverURL: String) {
        self.serverURL = serverURL.hasSuffix("/") ? String(serverURL.dropLast()) : serverURL
    }

    static func encodeStartBody(sessionID: String, participants: [String],
                                attrs: [String: String]) throws -> Data {
        var body: [String: Any] = ["session_id": sessionID, "participants": participants]
        if !attrs.isEmpty { body["attrs"] = attrs }
        return try JSONSerialization.data(withJSONObject: body, options: [.sortedKeys])
    }

    public func health() async throws {
        let url = URL(string: "\(serverURL)/admin/health")!
        let (_, resp) = try await URLSession.shared.data(from: url)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        guard code == 200 else { throw TransportError.http(code, "health") }
    }

    public func waitForHealth(timeout: TimeInterval = 15) async throws {
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            if (try? await health()) != nil { return }
            try await Task.sleep(nanoseconds: 200_000_000)
        }
        throw TransportError.timeout("server health")
    }

    public func startSession(sessionID: String, participants: [String],
                             attrs: [String: String] = [:]) async throws {
        var req = URLRequest(url: URL(string: "\(serverURL)/admin/sessions")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try Self.encodeStartBody(sessionID: sessionID, participants: participants, attrs: attrs)
        let (data, resp) = try await URLSession.shared.data(for: req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        guard code == 201 else { throw TransportError.http(code, String(decoding: data, as: UTF8.self)) }
    }

    public func getState(sessionID: String) async throws -> String {
        let url = URL(string: "\(serverURL)/admin/sessions/\(sessionID)")!
        let (data, _) = try await URLSession.shared.data(from: url)
        let obj = (try JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        return obj["state"] as? String ?? ""
    }

    public func pollUntilTerminal(sessionID: String, terminal: String,
                                  tries: Int = 40, interval: TimeInterval = 0.5) async throws -> String {
        var last = ""
        for _ in 0..<tries {
            last = (try? await getState(sessionID: sessionID)) ?? last
            if last.isEmpty || last == terminal { return last }
            try await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
        }
        return last
    }
}

File: clients/kotlin/ares-client/src/main/kotlin/ares/client/transport/AdminClient.kt
// SPDX-License-Identifier: Apache-2.0

package ares.client.transport

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * Thin REST wrapper for the ARES admin API.
 *
 * Mirrors Swift [AdminClient] semantics. All calls are synchronous and safe to
 * invoke from any thread.
 *
 * Endpoints used:
 *  - `GET  /admin/health`            → 200 on healthy
 *  - `POST /admin/sessions`          → 201 on success
 *  - `GET  /admin/sessions/{id}`     → `{"state":"..."}` JSON
 *
 * @param serverURL  base server URL, e.g. `http://localhost:8080` (trailing
 *                   slash is stripped automatically)
 */
class AdminClient(serverURL: String) {
    private val base    = serverURL.trimEnd('/')
    private val http    = OkHttpClient()
    private val stateRe = Regex("\"state\"\\s*:\\s*\"([^\"]*)\"")
    private val jsonMT  = "application/json".toMediaType()

    /**
     * Check whether the server is healthy (HTTP 200 on `/admin/health`).
     *
     * @return true if the server replied 200, false otherwise
     */
    fun health(): Boolean =
        runCatching {
            http.newCall(Request.Builder().url("$base/admin/health").build())
                .execute().use { it.isSuccessful }
        }.getOrDefault(false)

    /**
     * Block until the server reports healthy or [timeoutMs] elapses.
     *
     * Polls every 200 ms.
     *
     * @throws TransportException if the server does not become healthy in time
     */
    fun waitForHealth(timeoutMs: Long = 15_000) {
        val deadline = System.currentTimeMillis() + timeoutMs
        while (System.currentTimeMillis() < deadline) {
            if (health()) return
            Thread.sleep(200)
        }
        throw TransportException("AdminClient: server health timeout after ${timeoutMs}ms")
    }

    /**
     * Create a new ARES session via `POST /admin/sessions`.
     *
     * Request body: `{"session_id":<id>,"participants":[...],"attrs":{...}}`.
     * The `attrs` object is omitted when empty.
     *
     * @throws TransportException if the server does not return HTTP 201
     */
    fun startSession(
        sessionID: String,
        participants: List<String>,
        attrs: Map<String, String> = emptyMap(),
    ) {
        val sb = StringBuilder()
        sb.append("{\"session_id\":\"$sessionID\",\"participants\":[")
        sb.append(participants.joinToString(",") { "\"$it\"" })
        sb.append("]")
        if (attrs.isNotEmpty()) {
            sb.append(",\"attrs\":{")
            sb.append(attrs.entries.joinToString(",") { "\"${it.key}\":\"${it.value}\"" })
            sb.append("}")
        }
        sb.append("}")
        val body = sb.toString().toRequestBody(jsonMT)
        http.newCall(
            Request.Builder().url("$base/admin/sessions").post(body).build()
        ).execute().use { resp ->
            if (resp.code != 201) {
                throw TransportException(
                    "startSession: HTTP ${resp.code}: ${resp.body?.string()}"
                )
            }
        }
    }

    /**
     * Fetch the current state of session [sessionID] from `GET /admin/sessions/{id}`.
     *
     * @return state string (e.g. `"anon_gossip"`), or empty string on any error
     */
    fun getState(sessionID: String): String =
        runCatching {
            http.newCall(
                Request.Builder().url("$base/admin/sessions/$sessionID").build()
            ).execute().use { resp ->
                if (!resp.isSuccessful) return ""
                stateRe.find(resp.body?.string() ?: "")?.groupValues?.get(1) ?: ""
            }
        }.getOrDefault("")

    /**
     * Poll until the session reaches [terminal] state or [tries] is exhausted.
     *
     * @param terminal   target terminal state string
     * @param tries      maximum number of polling attempts (default 40)
     * @param intervalMs sleep between attempts in milliseconds (default 500)
     * @return last observed state (may not equal [terminal] if polling exhausted)
     */
    fun pollUntilTerminal(
        sessionID: String,
        terminal: String,
        tries: Int = 40,
        intervalMs: Long = 500,
    ): String {
        var last = ""
        repeat(tries) {
            last = runCatching { getState(sessionID) }.getOrDefault(last)
            if (last.isEmpty() || last == terminal) return last
            Thread.sleep(intervalMs)
        }
        return last
    }
}
